File: pset6/credit/credit.py
import sys

# "counter", shows current digit's position
count = 1
# final "checksum"
checksum = 0
# After the last step of the while loop last digit will be stored in n;
# The digit before it will be stored here in penultimateDigit
# We knowingly initialize it with invalid number since its unknown and can be determined after the first step
# and only if there is more than 1 digit in the number
penultimateDigit = -1

# Prompt user for the credit card number
# Prompt again if it's not an appropriate number
while True:
    n = input("Number: ")
    try:
        n = int(n)
        break
    except ValueError:
        continue

# Calculate the checksum
while n // 10 != 0:
    if count % 2 == 0:
        checksum += ((n % 10) * 2 % 10 + (n % 10) * 2 // 10)
    else:
        checksum += n % 10
    penultimateDigit = n % 10
    count += 1
    n //= 10

# Do the same for the last digit
if count % 2 == 0:
    checksum += (n * 2 % 10 + n * 2 // 10)
else:
    checksum += n

# Check if the checksum is correct
if checksum % 10 != 0:
    print("INVALID")
    sys.exit(0)

# Check for appropriate credit cards
if ((n == 3 and penultimateDigit == 4) or (n == 3 and penultimateDigit == 7)) and count == 15:
    print("AMEX")
elif ((n == 5 and penultimateDigit == 1)
      or (n == 5 and penultimateDigit == 2)
      or (n == 5 and penultimateDigit == 3)
      or (n == 5 and penultimateDigit == 4)
      or (n == 5 and penultimateDigit == 5)) and count == 16:
    print("MASTERCARD")
elif n == 4 and (count == 13 or count == 16):
    print("VISA")
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long n; // the credit card number
    int count = 1; // "counter", shows current digit's position
    int sum = 0; // final "checksum"
    int penultimateDigit = -1; // After the last step of the while loop last digit will be stored in n;
    // The digit before it will be stored here in penultimateDigit
    // We knowingly initialize it with invalid number since its unknown and can be determined after the first step
    // and only if there is more than 1 digit in the number

    n = get_long_long("Number: ");

    while (n / 10 != 0) // While there are left digits in the number
    {
        if (count % 2 == 0) // logic for digits on even positions
        {
            sum += ((n % 10) * 2 % 10 + (n % 10) * 2 / 10); // multiply by 2, get its digits' sum, add to the checksum

        }
        else // logic for digits on odd positions
        {
            sum += n % 10; // just add it to the checksum
        }
        penultimateDigit = n % 10; // store the digit as penultimateDigit
        count++; // increase the current digit's position
        n /= 10; // get rid off the proccessed digit
    }

    // Add the last number to the checksum
    if (count % 2 == 0) // if it is on even position
    {
        sum += (n * 2 % 10 + n * 2 / 10);
    }
    else // if it is on odd position
    {
        sum += n;
    }

    // if the checksum is invalid, print INVALID and exit the program
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Check for AmEx; AmEx cards start with 34 or 37 and has 15 digits in total
    if (((n == 3 && penultimateDigit == 4) || (n == 3 && penultimateDigit == 7)) && count == 15)
    {
        printf("AMEX\n");
    }
    // Check for MasterCard; MasterCard cards start with 51, 52, 53, 54 or 55 and has 16 digits in total
    else if (((n == 5 && penultimateDigit == 1)
              || (n == 5 && penultimateDigit == 2)
              || (n == 5 && penultimateDigit == 3)
              || (n == 5 && penultimateDigit == 4)
              || (n == 5 && penultimateDigit == 5)) && count == 16)
    {
        printf("MASTERCARD\n");
    }
    // Check for Visa; Visa cards start with 4 and has 13 or 16 digits in total
    else if (n == 4 && (count == 13 || count == 16))
    {
        printf("VISA\n");
    }
    // Otherwise the card number is invalid
    else
    {
        printf("INVALID\n");
    }
}